### ISSUE TITLE
Frequent intermittent connection failures

### ISSUE THREAD

**Issue by maroux** [python-poetry/poetry#3219]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: Debian GNU/Linux 10 (buster)
- **Poetry version**: 1.1.3
- **Link of a Gist with the contents of your pyproject.toml file**: https://gist.github.com/maroux/5bfc38375acd0e1c1a710fc0e62b8e93

## Issue
**Environment:** Docker debian slim buster image running in Google Cloud Platform Cloud Build
**Installation:** `poetry install -vvv --no-root --no-dev`
**Outcome:** Fails with networking error intermittently, but frequently enough that upgrade to poetry v1.1 is blocked on solving this issue.
**Notes:** Older version (v1.0.9) works fine but is slow because of old installer.
**Verbose logs:**
```
16:05:17 Step #2 - "build": Skipping virtualenv creation, as specified in config file.
16:05:17 Step #2 - "build": Installing dependencies from lock file
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build": Finding the necessary packages for the current system
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build": Package operations: 17 installs, 1 update, 0 removals, 11 skipped
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":   ? Installing certifi (2020.6.20): Skipped for the following reason: Already installed
16:05:17 Step #2 - "build":   ? Installing chardet (3.0.4): Skipped for the following reason: Already installed
16:05:17 Step #2 - "build":   ? Installing idna (2.10): Skipped for the following reason: Already installed
16:05:17 Step #2 - "build":   ? Installing markupsafe (1.1.1)
16:05:17 Step #2 - "build":   ? Installing pyparsing (2.4.7): Skipped for the following reason: Already installed
16:05:17 Step #2 - "build":   ? Installing pytz (2020.1): Skipped for the following reason: Already installed
16:05:17 Step #2 - "build":   ? Installing six (1.15.0): Skipped for the following reason: Already installed
16:05:17 Step #2 - "build":   ? Installing urllib3 (1.25.10): Skipped for the following reason: Already installed
16:05:17 Step #2 - "build":   ? Installing alabaster (0.7.12)
16:05:17 Step #2 - "build":   ? Installing babel (2.8.0)
16:05:17 Step #2 - "build":   ? Updating colorama (0.4.4 -> 0.4.3): Skipped for the following reason: Not needed for the current environment
16:05:17 Step #2 - "build":   ? Installing docutils (0.16): Skipped for the following reason: Already installed
16:05:17 Step #2 - "build":   ? Installing imagesize (1.2.0)
16:05:17 Step #2 - "build":   ? Installing jinja2 (2.11.2)
16:05:17 Step #2 - "build":   ? Installing packaging (20.4): Skipped for the following reason: Already installed
16:05:17 Step #2 - "build":   ? Updating pygments (2.7.1 -> 2.6.1)
16:05:17 Step #2 - "build":   ? Installing requests (2.24.0): Skipped for the following reason: Already installed
16:05:17 Step #2 - "build":   ? Installing snowballstemmer (2.0.0)
16:05:17 Step #2 - "build":   ? Installing sphinxcontrib-applehelp (1.0.1)
16:05:17 Step #2 - "build":   ? Installing sphinxcontrib-devhelp (1.0.1)
16:05:17 Step #2 - "build":   ? Installing sphinxcontrib-htmlhelp (1.0.2)
16:05:17 Step #2 - "build":   ? Installing sphinxcontrib-jsmath (1.0.1)
16:05:17 Step #2 - "build":   ? Installing sphinxcontrib-qthelp (1.0.2)
16:05:17 Step #2 - "build":   ? Installing sphinxcontrib-serializinghtml (1.1.3)
16:05:17 Step #2 - "build":   ? Installing commonmark (0.9.1)
16:05:17 Step #2 - "build":   ? Installing sphinx (3.1.2)
16:05:17 Step #2 - "build": Retrying HTTP request in 0.5 seconds.
16:05:17 Step #2 - "build":   ? Installing recommonmark (0.5.0)
16:05:17 Step #2 - "build":   ? Installing sphinx-autodoc-typehints (1.11.0)
16:05:17 Step #2 - "build":   ? Installing sphinx-rtd-theme (0.4.3)
16:05:17 Step #2 - "build": Retrying HTTP request in 0.5 seconds.
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":   Stack trace:
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":   13  ~/.poetry/lib/poetry/installation/executor.py:199 in _execute_operation
16:05:17 Step #2 - "build":        197? 
16:05:17 Step #2 - "build":        198?             try:
16:05:17 Step #2 - "build":      ? 199?                 result = self._do_execute_operation(operation)
16:05:17 Step #2 - "build":        200?             except EnvCommandError as e:
16:05:17 Step #2 - "build":        201?                 if e.e.returncode == -2:
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":   12  ~/.poetry/lib/poetry/installation/executor.py:273 in _do_execute_operation
16:05:17 Step #2 - "build":        271?             return 0
16:05:17 Step #2 - "build":        272? 
16:05:17 Step #2 - "build":      ? 273?         result = getattr(self, "_execute_{}".format(method))(operation)
16:05:17 Step #2 - "build":        274? 
16:05:17 Step #2 - "build":        275?         if result != 0:
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":   11  ~/.poetry/lib/poetry/installation/executor.py:408 in _execute_install
16:05:17 Step #2 - "build":        406? 
16:05:17 Step #2 - "build":        407?     def _execute_install(self, operation):  # type: (Install) -> None
16:05:17 Step #2 - "build":      ? 408?         return self._install(operation)
16:05:17 Step #2 - "build":        409? 
16:05:17 Step #2 - "build":        410?     def _execute_update(self, operation):  # type: (Update) -> None
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":   10  ~/.poetry/lib/poetry/installation/executor.py:434 in _install
16:05:17 Step #2 - "build":        432?             archive = self._download_link(operation, Link(package.source_url))
16:05:17 Step #2 - "build":        433?         else:
16:05:17 Step #2 - "build":      ? 434?             archive = self._download(operation)
16:05:17 Step #2 - "build":        435? 
16:05:17 Step #2 - "build":        436?         operation_message = self.get_operation_message(operation)
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":    9  ~/.poetry/lib/poetry/installation/executor.py:575 in _download
16:05:17 Step #2 - "build":        573? 
16:05:17 Step #2 - "build":        574?     def _download(self, operation):  # type: (Operation) -> Path
16:05:17 Step #2 - "build":      ? 575?         link = self._chooser.choose_for(operation.package)
16:05:17 Step #2 - "build":        576? 
16:05:17 Step #2 - "build":        577?         return self._download_link(operation, link)
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":    8  ~/.poetry/lib/poetry/installation/chooser.py:60 in choose_for
16:05:17 Step #2 - "build":         58?         """
16:05:17 Step #2 - "build":         59?         links = []
16:05:17 Step #2 - "build":      ?  60?         for link in self._get_links(package):
16:05:17 Step #2 - "build":         61?             if link.is_wheel and not Wheel(link.filename).is_supported_by_environment(
16:05:17 Step #2 - "build":         62?                 self._env
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":    7  ~/.poetry/lib/poetry/installation/chooser.py:94 in _get_links
16:05:17 Step #2 - "build":         92?             repository = self._pool.repository(package.source_reference)
16:05:17 Step #2 - "build":         93? 
16:05:17 Step #2 - "build":      ?  94?         links = repository.find_links_for_package(package)
16:05:17 Step #2 - "build":         95? 
16:05:17 Step #2 - "build":         96?         hashes = [f["hash"] for f in package.files]
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":    6  ~/.poetry/lib/poetry/repositories/pypi_repository.py:245 in find_links_for_package
16:05:17 Step #2 - "build":        243? 
16:05:17 Step #2 - "build":        244?     def find_links_for_package(self, package):
16:05:17 Step #2 - "build":      ? 245?         json_data = self._get("pypi/{}/{}/json".format(package.name, package.version))
16:05:17 Step #2 - "build":        246?         if json_data is None:
16:05:17 Step #2 - "build":        247?             return []
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":    5  ~/.poetry/lib/poetry/repositories/pypi_repository.py:319 in _get
16:05:17 Step #2 - "build":        317?     def _get(self, endpoint):  # type: (str) -> Union[dict, None]
16:05:17 Step #2 - "build":        318?         try:
16:05:17 Step #2 - "build":      ? 319?             json_response = self.session.get(self._base_url + endpoint)
16:05:17 Step #2 - "build":        320?         except requests.exceptions.TooManyRedirects:
16:05:17 Step #2 - "build":        321?             # Cache control redirect loop.
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":    4  ~/.poetry/lib/poetry/_vendor/py3.7/requests/sessions.py:543 in get
16:05:17 Step #2 - "build":        541? 
16:05:17 Step #2 - "build":        542?         kwargs.setdefault('allow_redirects', True)
16:05:17 Step #2 - "build":      ? 543?         return self.request('GET', url, **kwargs)
16:05:17 Step #2 - "build":        544? 
16:05:17 Step #2 - "build":        545?     def options(self, url, **kwargs):
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":    3  ~/.poetry/lib/poetry/_vendor/py3.7/requests/sessions.py:530 in request
16:05:17 Step #2 - "build":        528?         }
16:05:17 Step #2 - "build":        529?         send_kwargs.update(settings)
16:05:17 Step #2 - "build":      ? 530?         resp = self.send(prep, **send_kwargs)
16:05:17 Step #2 - "build":        531? 
16:05:17 Step #2 - "build":        532?         return resp
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":    2  ~/.poetry/lib/poetry/_vendor/py3.7/requests/sessions.py:643 in send
16:05:17 Step #2 - "build":        641? 
16:05:17 Step #2 - "build":        642?         # Send the request
16:05:17 Step #2 - "build":      ? 643?         r = adapter.send(request, **kwargs)
16:05:17 Step #2 - "build":        644? 
16:05:17 Step #2 - "build":        645?         # Total elapsed time of the request (approximately)
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":    1  ~/.poetry/lib/poetry/_vendor/py3.7/cachecontrol/adapter.py:53 in send
16:05:17 Step #2 - "build":         51?             request.headers.update(self.controller.conditional_headers(request))
16:05:17 Step #2 - "build":         52? 
16:05:17 Step #2 - "build":      ?  53?         resp = super(CacheControlAdapter, self).send(request, **kw)
16:05:17 Step #2 - "build":         54? 
16:05:17 Step #2 - "build":         55?         return resp
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":   ConnectionError
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":   ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
16:05:17 Step #2 - "build": 
16:05:17 Step #2 - "build":   at ~/.poetry/lib/poetry/_vendor/py3.7/requests/adapters.py:498 in send
16:05:17 Step #2 - "build":       494?                     low_conn.close()
16:05:17 Step #2 - "build":       495?                     raise
16:05:17 Step #2 - "build":       496? 
16:05:17 Step #2 - "build":       497?         except (ProtocolError, socket.error) as err:
16:05:17 Step #2 - "build":     ? 498?             raise ConnectionError(err, request=request)
16:05:17 Step #2 - "build":       499? 
16:05:17 Step #2 - "build":       500?         except MaxRetryError as e:
16:05:17 Step #2 - "build":       501?             if isinstance(e.reason, ConnectTimeoutError):
16:05:17 Step #2 - "build":       502?                 # TODO: Remove this in 3.0.0: see #2811
16:05:17 Step #2 - "build": 
```

**Comment by winpat** [comment]
I can confirm the same issue with Poetry 1.1.2. Every other install fail with a `ConnectionError`.

**Comment by abn** [comment]
Hmm, we saw this intermittently in our own CI a while ago, but was restricted to Cirrus CI FreeBSD instances (which runs on Google Cloud).

We might be able to do a bug fix to try recreate the session, but I suspect the issue might be at the process/scoket level.

@winpat what environment are you seeing your errors in?

I am really inclined to move http logic to `httpx` in the next minor to handle concurrent scenarios a bit better.

**Comment by winpat** [comment]
> @winpat what environment are you seeing your errors in?

I observed the problem in various GitLab CI pipelines (GitLab.com, not self hosted). If I remember correctly then GitLab also uses GCP.

**Comment by dariobig** [comment]
We are having the same issue with our CircleCI builds

**Comment by abn** [comment]
I am not entirely sure if this will fix the issue, since I am assuming the real issue here is some sort of networking layer throttling being hit. If some of you can please try the fix at #3230, it would be helpful to know if session reuse across thread exeuctions was the issue here.

### Using pipx
```sh
pipx install --suffix=@3230 'poetry @ git+https://github.com/python-poetry/poetry.git@refs/pull/3230/head'
```

### Using a container (podman | docker)
```sh
podman run --rm -i --entrypoint bash python:3.8 <<EOF
set -xe
pip install -q git+https://github.com/python-poetry/poetry.git@refs/pull/3230/head
poetry new foobar
pushd foobar
sed -i /pytest/d pyproject.toml
poetry add pycowsay
# commands to reproduce your build
EOF
```

**Comment by abn** [comment]
@maroux @winpat @dariobig can you please verify if the fix above resolves the issue for your environments. If it does I can get that into the next patch.

**Comment by winpat** [comment]
@abn Thanks for your help! I've added your fix to a couple of different projects. It will take a day or two until I can whether it's working or not.

**Comment by maroux** [comment]
> @maroux @winpat @dariobig can you please verify if the fix above resolves the issue for your environments. If it does I can get that into the next patch.

I've run it four times now and it's worked every time. I believe the PR does fix this issue.

**Comment by dariobig** [comment]
Unfortunately I can't test it in our environment, I've turned off "new
installer" instead and that fixed the issue for us.

**Comment by winpat** [comment]
> > @maroux @winpat @dariobig can you please verify if the fix above resolves the issue for your environments. If it does I can get that into the next patch.
> 
> I've run it four times now and it's worked every time. I believe the PR does fix this issue. 

I also ran it a couple of times without any issues so far 👍.

**Comment by sloev** [comment]
completely unusable because of this.
i get a finished build every 10 times from travis because i have 1 private package on gemfury.

*do i really have to live with this: ???*

```
Stack trace:
  13  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/poetry/installation/executor.py:199 in _execute_operation
       197│ 
       198│             try:
     → 199│                 result = self._do_execute_operation(operation)
       200│             except EnvCommandError as e:
       201│                 if e.e.returncode == -2:
  12  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/poetry/installation/executor.py:273 in _do_execute_operation
       271│             return 0
       272│ 
     → 273│         result = getattr(self, "_execute_{}".format(method))(operation)
       274│ 
       275│         if result != 0:
  11  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/poetry/installation/executor.py:408 in _execute_install
       406│ 
       407│     def _execute_install(self, operation):  # type: (Install) -> None
     → 408│         return self._install(operation)
       409│ 
       410│     def _execute_update(self, operation):  # type: (Update) -> None
  10  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/poetry/installation/executor.py:434 in _install
       432│             archive = self._download_link(operation, Link(package.source_url))
       433│         else:
     → 434│             archive = self._download(operation)
       435│ 
       436│         operation_message = self.get_operation_message(operation)
   9  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/poetry/installation/executor.py:575 in _download
       573│ 
       574│     def _download(self, operation):  # type: (Operation) -> Path
     → 575│         link = self._chooser.choose_for(operation.package)
       576│ 
       577│         return self._download_link(operation, link)
   8  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/poetry/installation/chooser.py:60 in choose_for
        58│         """
        59│         links = []
     →  60│         for link in self._get_links(package):
        61│             if link.is_wheel and not Wheel(link.filename).is_supported_by_environment(
        62│                 self._env
   7  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/poetry/installation/chooser.py:94 in _get_links
        92│             repository = self._pool.repository(package.source_reference)
        93│ 
     →  94│         links = repository.find_links_for_package(package)
        95│ 
        96│         hashes = [f["hash"] for f in package.files]
   6  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/poetry/repositories/pypi_repository.py:245 in find_links_for_package
       243│ 
       244│     def find_links_for_package(self, package):
     → 245│         json_data = self._get("pypi/{}/{}/json".format(package.name, package.version))
       246│         if json_data is None:
       247│             return []
   5  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/poetry/repositories/pypi_repository.py:319 in _get
       317│     def _get(self, endpoint):  # type: (str) -> Union[dict, None]
       318│         try:
     → 319│             json_response = self.session.get(self._base_url + endpoint)
       320│         except requests.exceptions.TooManyRedirects:
       321│             # Cache control redirect loop.
   4  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/requests/sessions.py:543 in get
       541│ 
       542│         kwargs.setdefault('allow_redirects', True)
     → 543│         return self.request('GET', url, **kwargs)
       544│ 
       545│     def options(self, url, **kwargs):
   3  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/requests/sessions.py:530 in request
       528│         }
       529│         send_kwargs.update(settings)
     → 530│         resp = self.send(prep, **send_kwargs)
       531│ 
       532│         return resp
   2  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/requests/sessions.py:643 in send
       641│ 
       642│         # Send the request
     → 643│         r = adapter.send(request, **kwargs)
       644│ 
       645│         # Total elapsed time of the request (approximately)
   1  ~/virtualenv/python3.7.4/lib/python3.7/site-packages/cachecontrol/adapter.py:53 in send
        51│             request.headers.update(self.controller.conditional_headers(request))
        52│ 
     →  53│         resp = super(CacheControlAdapter, self).send(request, **kw)
        54│ 
        55│         return resp
  ConnectionError
  ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
  at ~/virtualenv/python3.7.4/lib/python3.7/site-packages/requests/adapters.py:498 in send
      494│                     low_conn.close()
      495│                     raise
      496│ 
      497│         except (ProtocolError, socket.error) as err:
    → 498│             raise ConnectionError(err, request=request)
      499│ 
      500│         except MaxRetryError as e:
      501│             if isinstance(e.reason, ConnectTimeoutError):
      502│                 # TODO: Remove this in 3.0.0: see #2811
```

pyproject.toml:

```
[[tool.poetry.source]]
name = "gemfury"
url = "https://pypi.fury.io/SOME_COMPANY"

[tool.poetry]
name = "something_something"
version = "0.0.1"
description = ""
authors = []

[tool.poetry.dependencies]
python = "^3.7"
PRIVATE_STUFF= { version = "^6.6.6", source = "gemfury" }
OTHER_PRIVATE_STUFF= { version = "1.1.1", source = "gemfury" }
urlextract = "^0.14.0"
pandas = "^1.0.3"
emoji = "^0.5.4"
pendulum = "^2.1.0"
fastavro = "^0.23.4"
trustpilot-json-logging = ">=1.0.8"

[tool.poetry.dev-dependencies]
black = "^19.10b0"
pytest = "^5.4.1"
denv = "^2.0.1"
taskipy = "^1.2.1"


[build-system]
requires = ["poetry>=0.12"]
build-backend = "poetry.masonry.api"
```

**Comment by sloev** [comment]
my build just got thru travis. *TOOK 11 TIMES* for poetry to *NOT TIMEOUT*

**Comment by DustinMoriarty** [comment]
I am actually seeing a similar issue during test set up which only occurs for Python 3.6 with FreeBSD for the poetry PR #3259.  

```
12  .tox/py/lib/python3.6/site-packages/poetry/installation/executor.py:273 in _do_execute_operation
      result = getattr(self, "_execute_{}".format(method))(operation)

  11  .tox/py/lib/python3.6/site-packages/poetry/installation/executor.py:408 in _execute_install
      return self._install(operation)

  10  .tox/py/lib/python3.6/site-packages/poetry/installation/executor.py:434 in _install
      archive = self._download(operation)

   9  .tox/py/lib/python3.6/site-packages/poetry/installation/executor.py:575 in _download
      link = self._chooser.choose_for(operation.package)

   8  .tox/py/lib/python3.6/site-packages/poetry/installation/chooser.py:60 in choose_for
      for link in self._get_links(package):

   7  .tox/py/lib/python3.6/site-packages/poetry/installation/chooser.py:94 in _get_links
      links = repository.find_links_for_package(package)

   6  .tox/py/lib/python3.6/site-packages/poetry/repositories/pypi_repository.py:245 in find_links_for_package
      json_data = self._get("pypi/{}/{}/json".format(package.name, package.version))

   5  .tox/py/lib/python3.6/site-packages/poetry/repositories/pypi_repository.py:319 in _get
      json_response = self.session.get(self._base_url + endpoint)

   4  .tox/py/lib/python3.6/site-packages/requests/sessions.py:543 in get
      return self.request('GET', url, **kwargs)

   3  .tox/py/lib/python3.6/site-packages/requests/sessions.py:530 in request
      resp = self.send(prep, **send_kwargs)

   2  .tox/py/lib/python3.6/site-packages/requests/sessions.py:643 in send
      r = adapter.send(request, **kwargs)

   1  .tox/py/lib/python3.6/site-packages/cachecontrol/adapter.py:53 in send
      resp = super(CacheControlAdapter, self).send(request, **kw)

  ConnectionError

  ('Connection aborted.', ConnectionResetError(54, 'Connection reset by peer'))

  at .tox/py/lib/python3.6/site-packages/requests/adapters.py:498 in send
      494│                     low_conn.close()
      495│                     raise
      496│ 
      497│         except (ProtocolError, socket.error) as err:
    → 498│             raise ConnectionError(err, request=request)
      499│ 
      500│         except MaxRetryError as e:
      501│             if isinstance(e.reason, ConnectTimeoutError):
      502│                 # TODO: Remove this in 3.0.0: see #2811

ERROR: InvocationError for command /tmp/cirrus-ci-build/.tox/py/bin/poetry install -vv --no-root (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   py: commands failed
```

**Comment by sloev** [comment]
when is this gonna be fixed? this morning: again 9 consecutive travis builds before i got thru my poetry install step. poetry version: 1.1.2

**Comment by abn** [comment]
@sloev as I mentioned on the pull request (https://github.com/python-poetry/poetry/pull/3230#issuecomment-713551777), this is an open source community project, not a product you have a paid support for. This is maintained by a small group of volunteers who spend their spare time on this. You demanding issues be fixed and spamming issue threads is not going to make anything go any faster. While we can appreciate that this is causing issues with your workflow, we have mentioned in the [release post](https://python-poetry.org/blog/announcing-poetry-1-1-0.html#brand-new-installer) how to disable the experimental installer to unblock yourself. Please keep the conversation constructive and civil.

**Comment by abn** [comment]
@DustinMoriarty yeah; we noticed that. This should go away once `1.1.4` is out. Hopefully the fix in #3230 is sufficient.

**Comment by abn** [comment]
This was resolved with `1.1.4`.

**Comment by maroux** [comment]
uh-oh just saw this now on 1.1.4:
```
13:08:47 Step #2 - "build":   ChunkedEncodingError
13:08:47 Step #2 - "build": 
13:08:47 Step #2 - "build":   ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
13:08:47 Step #2 - "build": 
13:08:47 Step #2 - "build":   at ~/.poetry/lib/poetry/_vendor/py3.7/requests/models.py:754 in generate
13:08:47 Step #2 - "build":       750?                 try:
13:08:47 Step #2 - "build":       751?                     for chunk in self.raw.stream(chunk_size, decode_content=True):
13:08:47 Step #2 - "build":       752?                         yield chunk
13:08:47 Step #2 - "build":       753?                 except ProtocolError as e:
13:08:47 Step #2 - "build":     ? 754?                     raise ChunkedEncodingError(e)
13:08:47 Step #2 - "build":       755?                 except DecodeError as e:
13:08:47 Step #2 - "build":       756?                     raise ContentDecodingError(e)
13:08:47 Step #2 - "build":       757?                 except ReadTimeoutError as e:
13:08:47 Step #2 - "build":       758?                     raise ConnectionError(e)
13:08:47 Step #2 - "build": 
13:08:47 Step #2 - "build":   ? Installing snowballstemmer (2.0.0)
13:08:47 Step #2 - "build":   ? Installing sphinxcontrib-applehelp (1.0.1)
13:09:01 ERROR: (gcloud.builds.submit) build 17e8547a-8452-4010-a146-81c7f21a4309 completed with status "FAILURE"
```
might be unrelated?

**Comment by Korijn** [comment]
We're still seeing exactly this problem on our CI builds (azure pipelines) with 1.1.4.

**Comment by SimonBiggs** [comment]
> Since there's a pretty big list of packages to install, this happens almost every build. We're pinning to poetry < 1.1 for now as a workaround.

As per https://python-poetry.org/blog/announcing-poetry-1-1-0.html#brand-new-installer adding something like the following to your CI builds might work:

https://github.com/pymedphys/pymedphys/blob/765f6229105f221c3dd75ff9927e62dfde1b3972/.github/workflows/test-and-deploy.yml#L154-L156

**Comment by prnvkbr** [comment]
Unfortunately, still experiencing this on `poetry-1.1.5`. Has anyone else experienced this issue lately?

```
...
#22 177.1 Package operations: 47 installs, 0 updates, 0 removals
#22 177.1 
#22 177.1   • Installing numpy (1.20.1)
#22 177.1   • Installing typing-extensions (3.7.4.3)
#22 184.8   • Installing markupsafe (1.1.1)
#22 184.8   • Installing pillow (8.1.2)
#22 184.8   • Installing six (1.15.0)
#22 184.8   • Installing smmap (3.0.5)
#22 184.8   • Installing torch (1.7.1)
#22 195.9 
#22 195.9   ChunkedEncodingError
#22 195.9 
#22 195.9   ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
#22 195.9 
#22 195.9   at /opt/conda/lib/python3.8/site-packages/requests/models.py:754 in generate
#22 196.1       750│                 try:
#22 196.1       751│                     for chunk in self.raw.stream(chunk_size, decode_content=True):
#22 196.1       752│                         yield chunk
#22 196.1       753│                 except ProtocolError as e:
#22 196.1     → 754│                     raise ChunkedEncodingError(e)
#22 196.1       755│                 except DecodeError as e:
#22 196.1       756│                     raise ContentDecodingError(e)
#22 196.1       757│                 except ReadTimeoutError as e:
#22 196.1       758│                     raise ConnectionError(e)
#22 196.1 
------
Dockerfile:21
--------------------
  19 |     RUN pip install --upgrade pip && pip --no-cache-dir install poetry
  20 |     COPY ./pyproject.toml .
  21 | >>> RUN poetry install --verbose
```
Base Docker image: `continuumio/miniconda3`. 

Any workarounds?

**Comment by taranlu-houzz** [comment]
I'm also getting a related error intermittently on `1.1.5`:
```
[...]

Using version ^6.2.2 for pytest

Updating dependencies
Resolving dependencies... (3.0s)

Writing lock file

Package operations: 8 installs, 0 updates, 0 removals

  • Installing pyparsing (2.4.7)
  • Installing attrs (20.3.0)
  • Installing iniconfig (1.1.1)
  • Installing packaging (20.9): Failed

  ConnectionError

  ('Connection aborted.', ConnectionResetError(54, 'Connection reset by peer'))

  at ~/.poetry/lib/poetry/_vendor/py3.9/requests/adapters.py:498 in send
      494│                     low_conn.close()
      495│                     raise
      496│
      497│         except (ProtocolError, socket.error) as err:
    → 498│             raise ConnectionError(err, request=request)
      499│
      500│         except MaxRetryError as e:
      501│             if isinstance(e.reason, ConnectTimeoutError):
      502│                 # TODO: Remove this in 3.0.0: see #2811

  • Installing pluggy (0.13.1)
  • Installing py (1.10.0)
  • Installing toml (0.10.2)

Failed to add packages, reverting the pyproject.toml file to its original content.
```

**Comment by taranlu-houzz** [comment]
@SimonBiggs Is there a different issue that I should follow to track updates on this? The connection error issues are making it so that I have to disable the new installer for the time being.

**Comment by btjones-me** [comment]
Hi folks, I'm also having what I think is the same/a related problem failing on my builds, maybe 50% of calls to a private package repository (Azure Artifacts Feed) are failing with the below. Interested to know if anybody else has a workaround/ solved.

```
$  poetry install
$ make test-all
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.7.10/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.7.10/x64/lib
No suitable keyring backends were found
Using a plaintext file to store and retrieve credentials
Creating virtualenv ***_-py3.7 in /home/runner/.cache/pypoetry/virtualenvs
Updating dependencies
Resolving dependencies...

  ConnectionError

  ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))

  at ~/.poetry/lib/poetry/_vendor/py3.7/requests/adapters.py:498 in send
      494│                     low_conn.close()
      495│                     raise
      496│ 
      497│         except (ProtocolError, socket.error) as err:
    → 498│             raise ConnectionError(err, request=request)
      499│ 
      500│         except MaxRetryError as e:
      501│             if isinstance(e.reason, ConnectTimeoutError):
      502│                 # TODO: Remove this in 3.0.0: see #2811
      ```

**Comment by danielgafni** [comment]
Hey guys!

I've encountered this issue on a machine with a lot of CPUs (128). 
Looks like `poetry install` is basically DDOSing PYPI if you have a lot of cores.

Temporal solution: `poetry config installer.parallel false`.
Docker solution: limit the number of CPUs available to `docker build`. (`--cpuset-cpus <num>`).

Perhaps a `poetry config installer.max_cpus` settings would be a good final solution.

**Comment by Korijn** [comment]
How about max connections, instead of max cpus. And  the default should be below pypi's rate limits.

**Comment by abn** [comment]
@danielgafni can you open a new issue for this? The workers need to be configurable. We definitely are a bit greedy, but then again we I suspct what we need to connection pooling for repository sessions so installations can continue faster if all dependencies are already cached for example.

https://github.com/python-poetry/poetry/blob/c6b19b348d86390135e16ac17e426e5c41202b16/poetry/installation/executor.py#L75

**Comment by cposada23** [comment]
Hello I keep getting this 

```
 ('Connection aborted.', TimeoutError(10060, 'A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond', None, 10060, None))

  at ~\AppData\Roaming\pypoetry\venv\lib\site-packages\requests\adapters.py:498 in send
      494│                     low_conn.close()
      495│                     raise
      496│
      497│         except (ProtocolError, socket.error) as err:
    → 498│             raise ConnectionError(err, request=request)
      499│
      500│         except MaxRetryError as e:
      501│             if isinstance(e.reason, ConnectTimeoutError):
      502│                 # TODO: Remove this in 3.0.0: see #2811
```
When I run poetry Install, have anyone find a solution ?

**Comment by danielgafni** [comment]
hey @abn ! I'm sorry, I totally forgot about this thread. 

I've found a [PR](https://github.com/python-poetry/poetry/pull/3516) that implements exactly this, it looks like it's almost ready.

**Comment by rbudnar** [comment]
We're experiencing this issue as well very consistently in our repositories that include a link to our private Azure package feed. We've tried the solution posted [above](https://github.com/python-poetry/poetry/issues/3219#issuecomment-955496032) to no avail. 

We've noticed that when we remove our private packages and package feed from our list, the problem goes away immediately; also dependency resolution is much much faster.

**Comment by mbelang** [comment]
I have the same issue as @rbudnar with 1.1.12 with Gemfury. The failures are really uneven though sometimes it fails right away, sometimes it takes 30 secs.

**Comment by gshpychka** [comment]
@abn I'm also still running into this, even with `installer.parallel false`, when running on AWS CodeBuild.

**Comment by abn** [comment]
@gshpychka @mbelang  can you try with poetry@master maybe? If that does not work, maybe raise a new issue?

**Comment by roscioli** [comment]
+1 -- happening on poetry 1.1.13 in a docker container :(

**Comment by taranlu-houzz** [comment]
I am also still experiencing this. I have not had time to investigate further, but I am wondering if it might have to do with using a private pypi repo as an additional source? I don't seem to run into this issue when I am working on packages that don't have the `tool.poetry.source` entry. Seems like maybe it could be a combination of the private pypi server, VPN settings, and potentially underpowered hardware for the pypi server, since the issues are due to timeouts...

**Comment by roscioli** [comment]
My hunch is that it also may have to do with internet speed. I'm succeeding at 330 Mbps but my colleague is failing at 27 Mbps. I think perhaps the retries need to be more forgiving?

**Comment by danielgafni** [comment]
Just fyi if you haven't tried it yet: setting `poetry config installer.parallel false` fixes the issue for me in 100% of the cases. We are running a lot of CI with private repos and this issue never occurs for us (with the correct setting). It may occur if installing Poetry projects on machines with a lot of CPUs (without the setting).

**Comment by invokermain** [comment]
@danielgafni I've added that to our pipelines today, shame that it makes the pipelines dramatically slower (and the first pipeline we ran after it failed with the error). But hopefully it will help a bit.

Out of interest are your private repos set as 'secondary = True' or not?

**Comment by danielgafni** [comment]
@invokermain yes, it's set to true

**Comment by invokermain** [comment]
Yeah a few days later and we are still getting this issue very regularly, even with `poetry config installer.parallel false`. I think one of the underlying issues that means it's more frequent in CI/CD is the fact that you are less likely to have caching enabled. I.e. for us there is no caching and therefore the installs are much more spammy (especially if running multiple jobs, e.g. test/quality/security, in parallel).

**Comment by invokermain** [comment]
Hmmm looks like this might be resolved on 1.2, so 🤞 for a release soon!

**Comment by illyakaynov** [comment]
Still experiencing this problem on 1.2.2

**Comment by mohghaderi** [comment]
Still experiencing this problem on 1.4.x
`poetry config installer.parallel false` works almost always

`poetry config installer.max-workers 4` still fails most of the time
`poetry config installer.max-workers 2` works
 
I think there should be a problem with the way parallel works.

**Comment by harsh-divecha** [comment]
> n3.7/ssl.py:929 in read
>         925│         if self._sslobj is None:
>         926│             raise ValueError("Read on closed or unwrapped SSL socket.")
>         927│         try:
>         928│             if buffer is not None:
>      →  929│                 return self._sslobj.read(len, buffer)
>         930│             else:
>         931│                 return self._sslobj.read(len)
>         932│         except SSLError as x:
>         933│             if x.args[0] == SSL_ERROR_EOF and self.suppress_ragged_eofs:
>  
>  The following error occurred when trying to handle this error:

Facing exactly the same.

**Comment by ezorita** [comment]
Same behavior with poetry `1.5.1`, intermittent failures when installing in docker containers or while building images. @abn would you consider reopening this issue to investigate? Thanks!

**Comment by gehirndienst** [comment]
I confirm the same issue with poetry `1.6.1`, the only thing that brings stability (in a turtle-speed way though) is `poetry config installer.parallel false`

**Comment by maronuu** [comment]
I also confirmed the same issue with poetry `1.6.1`, even when `poetry config installer.parallel false` is set.
I only encountered the error while building a docker image from Dockerfile on MacOS. In my case, various package (non-deterministic) other than nvidia-packages failed.
